src/server/internalServer.cpp

```diff
@@ -95,35 +95,200 @@ inline std::string normalizeRootUrl(std::string rootUrl)
+    } catch (...) {}
+    return filter;
+}
 
-  return defaultVal;
+template<class T>
+std::vector<T> subrange(const std::vector<T>& v, size_t s, size_t n)
+{
+  const size_t e = std::min(v.size(), s+n);
+  return std::vector<T>(v.begin()+std::min(v.size(), s), v.begin()+e);
 }
+
+std::string renderUrl(const std::string& root, const std::string& urlTemplate)
+{
+  MustacheData data;
+  data.set("root", root);
+  auto url = kainjow::mustache::mustache(urlTemplate).render(data);
+  if ( url.back() == '\n' )
+    url.pop_back();
+  return url;
+}
+
+std::string makeFulltextSearchSuggestion(const std::string& lang, const std::string& queryString)
+{
+  return i18n::expandParameterizedString(lang, "suggest-full-text-search",
+               {
+                  {"SEARCH_TERMS", queryString}
+               }
+         );
+}
+
+ParameterizedMessage noSuchBookErrorMsg(const std::string& bookName)
+{
+  return ParameterizedMessage("no-such-book", { {"BOOK_NAME", bookName} });
+}
+
+ParameterizedMessage invalidRawAccessMsg(const std::string& dt)
+{
+  return ParameterizedMessage("invalid-raw-data-type", { {"DATATYPE", dt} });
+}
+
+ParameterizedMessage noValueForArgMsg(const std::string& argument)
+{
+  return ParameterizedMessage("no-value-for-arg", { {"ARGUMENT", argument} });
+}
+
+ParameterizedMessage rawEntryNotFoundMsg(const std::string& dt, const std::string& entry)
+{
+  return ParameterizedMessage("raw-entry-not-found",
+                              {
+                                {"DATATYPE", dt},
+                                {"ENTRY", entry},
+                              }
+  );
+}
+
+ParameterizedMessage tooManyBooksMsg(size_t nbBooks, size_t limit)
+{
+  return ParameterizedMessage("too-many-books",
+                              {
+                                {"NB_BOOKS", nbBooks},
+                                {"LIMIT", limit},
+                              }
+  );
+}
+
+ParameterizedMessage nonParameterizedMessage(const std::string& msgId)
+{
+  const ParameterizedMessage::Parameters noParams;
+  return ParameterizedMessage(msgId, noParams);
+}
+
+struct Error : public std::runtime_error {
+  explicit Error(const ParameterizedMessage& message)
+    : std::runtime_error("Error while handling request"),
+      _message(message)
+  {}
+
+  const ParameterizedMessage& message() const
+  {
+    return _message;
+  }
+
+  const ParameterizedMessage _message;
+};
+
+void checkBookNumber(const Library::BookIdSet& bookIds, size_t limit) {
+  if (bookIds.empty()) {
+    throw Error(nonParameterizedMessage("no-book-found"));
+  }
+  if (limit > 0 && bookIds.size() > limit) {
+    throw Error(tooManyBooksMsg(bookIds.size(), limit));
+  }
+}
+
 } // unnamed namespace
 
-SearchInfo::SearchInfo(const std::string& pattern)
-  : pattern(pattern),
-    geoQuery()
-{}
+std::pair<std::string, Library::BookIdSet> InternalServer::selectBooks(const RequestContext& request, const std::string& prefix) const
```

Comment:
It looks like we don't need `prefix` as a parameter for `selectBooks()` and `getSearchInfo()` (since the latter is called only with a prefix of "books" while being the only caller of the former). If `prefix` is justified later, let's add it when we really need it.